### PR TITLE
[mypy] Fix class name is not defined error

### PR DESCRIPTION
**Pull request by mayitbeegh**
Signed-off-by: Sean Lin <sean@union.ai>

# TL;DR
Fix class name is not defined error in mypy. Some errors are fixed by importing necessary modules and placing forward reference with normal reference; Others are fixed by introducing forward reference to avoid circular dependency https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1428

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/types/schema/types_pandas.py [hunk 1]
```diff
@@ -10,14 +10,17 @@
 from flytekit.plugins import pandas
```
Comment (wild-endeavor) [inline flytekit/types/schema/types_pandas.py:10]: @kumare3 can you take a look at this one line please?  We should just swap it out right?

File: flytekit/types/schema/types_pandas.py [hunk 2]
```diff
@@ -10,14 +10,17 @@
...
 from flytekit.types.schema import LocalIOSchemaReader, LocalIOSchemaWriter, SchemaEngine, SchemaFormat, SchemaHandler
 
+if typing.TYPE_CHECKING:
+    import pandas as _pandas
```
Comment (wild-endeavor) [inline flytekit/types/schema/types_pandas.py:14]: pending the comment above from ketan, i think we can just move this up.

File: tests/flytekit/unit/core/test_flyte_directory.py [hunk 3]
```diff
@@ -140,8 +140,10 @@ def my_wf() -> FlyteDirectory:
     files = os.listdir(wfd.path)
     assert len(files) == 5
 
+    csv = str
+
     @task
-    def t2(in1: FlyteDirectory["csv"]) -> int:
+    def t2(in1: FlyteDirectory[csv]) -> int:
```
Comment (wild-endeavor) [inline tests/flytekit/unit/core/test_flyte_directory.py:146]: It's a hack but could we use this style instead please https://github.com/flyteorg/flytekit/blob/b8a3c7d34e6b19722a6a968dc05506ad1eb26912/flytekit/types/file/__init__.py#L60?  It's a hack because it just so happens that the string representation is the string that you pass in, but whatever. generics in general are pretty hacky in python.

**Review by wild-endeavor — COMMENTED** [review]
just a few comments.  thank you!!!

File: flytekit/types/schema/types_pandas.py (hunk 1 above)
Comment (kumare3) [reply flytekit/types/schema/types_pandas.py:10]: yup, flytekit has a direct dependency on pandas.

File: tests/flytekit/unit/core/test_flyte_directory.py (hunk 3 above)
Comment (kumare3) [reply tests/flytekit/unit/core/test_flyte_directory.py:146]: maybe there is an alternative?

**Review by wild-endeavor — APPROVED** [review]
